blog: EmberData Polaris Update (#1272)
ember-learn/ember-blog: content/ember-data-polaris-2023-04-11.md

---
title: EmberData Polaris Update
authors:

>   - chris-thoburn

Reviewer: You'll want to add more authors here as appropriate, or change the "we" statements throughout the article.

date: 2023-04-11T00:00:00.000Z
tags:
  - '2023'
  - ember-data
  - ember-polaris

---

This week, in coordination with the broader Ember project, EmberData released 4.12, its final
4.x version, and began iterating towards the first release of the 5.x series.

## A wrap on 4.x

This final release of the 4.x cycle introduced several new concepts to support a new paradigm for managing [Requests]() and new capabilities for [Caching Data]().

The primary focus of these changes was to support the concept of a [Document]() as a cache primitive and to complete the work to ensure that the [Resource]() concept is handled opaquely.

> Collectively, these changes allow applications the capability of building highly advanced request and cache capabilities overtop of familiar EmberData paradigms. We suggest reading the respective RFCs for [RequestManager]() and [Cache]() to gain a fuller appreciation of the motivations and capabilities they bring.

Reviewer: ```suggestion
Collectively, these changes allow applications the capability of building highly advanced request and cache capabilities on top of of familiar EmberData paradigms. We suggest reading the respective RFCs for [RequestManager]() and [Cache]() to gain a fuller appreciation of the motivations and capabilities they bring.
```

While these new primitives may not at first feel like a substantial shift, over the course of 5.x as the new request story is polished we expect for the recommended experience when using EmberData to significantly evolve, with that evolution driven by these changes.

## Looking ahead to 5.x

Throughout the 5.x cycle we expect to introduce two important new defaults while deprecating an old stalwart.

For nearly 17 years, [Model](https://github.com/sproutcore/sproutcore/commit/f6248b1650a688a401cc6eea135fbe983e20cd12#diff-011979c89114a908391f35c2053dc2ba84da4d331cc97730039b2b2da623ffee) has been a foundational primitive around which EmberData was understood. Since those earliest days, the language and ecosystem have evolved, the kinds of applications we build has evolved, and the patterns by which we access and mutate data have evolved. While Model has undergone small amounts of evolution in syntax, its underlying patterns have remained unevolved.

> In more recent years, various 3rd party attempts have been made to address some of the shortcomings of the Model paradigm. [ember-m3]() explored what it might mean to have models be schema-driven. [ember-data-model-fragments]() offered a workaround for deep-tracking. [ember-data-storefront]() offered alternative data access patterns to simplify the mental model of working with asyncronous edges in relational data. [ember-data-changetracker](), [ember-changesets](), and [ember-buffered-proxy]() offered mechanisms for attempting to streamline mutation flows and make changes more easily discardable.

Reviewer: ```suggestion
In more recent years, various 3rd party attempts have been made to address some of the shortcomings of the Model paradigm. [ember-m3]() explored what it might mean to have schema-driven models. [ember-data-model-fragments]() offered a deep-tracking workaround. [ember-data-storefront]() offered alternative data access patterns to simplify the mental model of asyncronous edges in relational data. [ember-data-changetracker](), [ember-changesets](), and [ember-buffered-proxy]() offered mechanisms for streamlined mutation flows and more easily discardable changes.
```

> While we appreciate each of these addons for taking on a challenging aspect of EmberData and solving real needs of our consumers, the nature of many of these features having been built overtop a shaky foundation and against a shifting target of private and intimate API contracts has meant that applications using these addons have faced unexpectedly large challenges when upgrading their version of EmberData.

Reviewer: ```suggestion
Although we appreciate each of these addons for tackling challenging aspects of EmberData and addressing the genuine needs of our consumers, many of these features were built on an unstable foundation and rely on constantly changing private and intimate API contracts. Consequently, applications using these addons have encountered significant difficulties when attempting to upgrade their version of EmberData.
```

Suggestion to break up a long sentence.

---
title: EmberData Polaris Update
authors:
  - chris-thoburn
  - the-ember-data-team

> date: 2023-04-11T00:00:00.000Z

Reviewer 2: ```suggestion
date: 2023-04-12T00:00:00.000Z
```

tags:
  - '2023'
  - ember-data
  - ember-polaris

---

This week, in coordination with the broader Ember project, EmberData released 4.12, its final
4.x version, and began iterating towards the first release of the 5.x series.

## A wrap on 4.x

This final release of the 4.x cycle introduced several new concepts to support a new paradigm for managing [Requests](https://github.com/emberjs/data/tree/v4.12.0/packages/request#readme) and new capabilities for [Caching Data](https://api.emberjs.com/ember-data/release/classes/%3CInterface%3E%20Cache).

> The primary focus of these changes was to support caching **Documents** and to ensure that **Resources** are handled opaquely. Loosely defined: A Document encompasses the full response returned by an API request, and can typically be identified uniquely by url. Documents may contain many resources or references to resources, where a resource is defined as some data which your application or the API treats as uniquely identitifiable.

Reviewer 2: ```suggestion
The primary focus of these changes was to support caching **Documents** and to ensure that **Resources** are handled opaquely. Loosely defined: A Document encompasses the full response returned by an API request, and can typically be identified uniquely by url. Documents may contain many resources or references to resources, where a resource is defined as some data which your application or the API treats as uniquely identifiable.
```

> > *Note:* Historically, Model has mapped to a resource "type" and serves to present the data for a resource. It may be useful to think about resources as the backing data for a Model; however, resources can be far more generic than this and could even be anything. This includes raw strings, html, xml, buffers, images, blobs, and streams!

Reviewer 2: I like the way this looks as a block because of the styling, but the block styling also sorta makes it look like we're quoting from somewhere else because of the quote icon it employs. Not really a blocker, just something to consider.

Author: yeah I thought the same. It's definitely an aside, not sure if we have a good way to easily style that ...

Reviewer 2: Was looking for the styling of `EsNote` from the ember-styleguide, but it seems broken here in the blog. I'll have to look into that.

While these new primitives may not at first feel like a substantial shift, over the course of 5.x as the new request story is polished we expect for the recommended experience when using EmberData to significantly evolve, with that evolution driven by these changes.

## Looking ahead to 5.x

Throughout the 5.x cycle we expect to introduce two important new defaults while deprecating an old stalwart.

For nearly 17 years, [Model](https://github.com/sproutcore/sproutcore/commit/f6248b1650a688a401cc6eea135fbe983e20cd12#diff-011979c89114a908391f35c2053dc2ba84da4d331cc97730039b2b2da623ffee) has been a foundational primitive around which EmberData was understood. Since those earliest days, the language and ecosystem have evolved, the kinds of applications we build has evolved, and the patterns by which we access and mutate data have evolved. While Model has undergone small amounts of evolution in syntax, its underlying patterns have remained unevolved.

> In more recent years, various 3rd party attempts have been made to address some of the shortcomings of the Model paradigm. [ember-m3](https://github.com/hjdivad/ember-m3) explored what it might mean to have schema-driven models. [ember-data-model-fragments](https://github.com/adopted-ember-addons/ember-data-model-fragments) offered a deep-tracking workaround. [ember-data-storefront](https://github.com/embermap/ember-data-storefront) offered alternative data access patterns to simplify the mental model of asyncronous edges in relational data. [ember-data-changetracker](https://github.com/danielspaniel/ember-data-change-tracker), [ember-changeset](https://github.com/poteto/ember-changeset), and [ember-buffered-proxy](https://github.com/yapplabs/ember-buffered-proxy) offered mechanisms for streamlined mutation flows and more easily discardable changes.

Reviewer 2: ```suggestion
In more recent years, various 3rd party attempts have been made to address some of the shortcomings of the Model paradigm. [ember-m3](https://github.com/hjdivad/ember-m3) explored what it might mean to have schema-driven models. [ember-data-model-fragments](https://github.com/adopted-ember-addons/ember-data-model-fragments) offered a deep-tracking workaround. [ember-data-storefront](https://github.com/embermap/ember-data-storefront) offered alternative data access patterns to simplify the mental model of asynchronous edges in relational data. [ember-data-changetracker](https://github.com/danielspaniel/ember-data-change-tracker), [ember-changeset](https://github.com/poteto/ember-changeset), and [ember-buffered-proxy](https://github.com/yapplabs/ember-buffered-proxy) offered mechanisms for streamlined mutation flows and more easily discardable changes.
```

Although we appreciate each of these addons for tackling challenging aspects of EmberData and addressing the genuine needs of our consumers, many of these features were built on an unstable foundation and rely on constantly changing private and intimate API contracts. Consequently, applications using these addons have encountered significant difficulties when attempting to upgrade their version of EmberData.

> So throughout the 3.x and 4.x cycles, we began evolving a new set up public APIs on top which we could support these ideas without compromise, and plotting to bring ideas explored by each of these libraries into the core recommended experience of EmberData itself. RequestManager and Cache are the result of these efforts.

Reviewer 2: ```suggestion
So throughout the 3.x and 4.x cycles, we began evolving a new set of public APIs on top which we could support these ideas without compromise, and plotting to bring ideas explored by each of these libraries into the core recommended experience of EmberData itself. RequestManager and Cache are the result of these efforts.
```

This means (for instance) that the migration path for an application using [ember-data-model-fragments](https://github.com/adopted-ember-addons/ember-data-model-fragments) is to help that library become compatible with 4.12, update their application to 4.12, and begin incrementally migrating from Model+ModelFragments to the new presentation class which will offer deep-tracking out of the box.

For [ember-data-model-fragments](https://github.com/adopted-ember-addons/ember-data-model-fragments) specifically, we commit to providing extensive time to assist in adding support for 4.12 to the current 3.28+ branch that some teams have been using today. In this way, we hope to give even those teams facing hurdles in the jump to 4.0 a carrot and a mechanism to leap quickly to 5.x.

> ## 4.12 will become a Special LTS

Reviewer 2: ```suggestion
## EmberData 4.12 will become a Special LTS
```

> In addition to the committments above, we are planning on declaring 4.12 as a special LTS release.

Reviewer 2: ```suggestion
In addition to the commitments above, we are planning on declaring EmberData 4.12 as a special LTS release.
```

---
title: EmberData Polaris Update
authors:
  - chris-thoburn
  - the-ember-data-team

> date: 2023-04-12T00:00:00.000Z

Reviewer 2: ```suggestion
date: 2023-04-14T00:00:00.000Z
```

Reviewer 2: You might want to remove the date from the file name.

Author: > You might want to remove the date from the file name.

the other file names all use a date though, and better for consistency. Happy to update the date if we care strongly that it matches the publish vs the create data.
